JIT build crashes on assertions (since 3.15): `Python/pystate.c:925: interpreter_clear: Assertion 'cold->vm_data.valid' failed.`

# Crash report

### What happened?

```
./configure --enable-experimental-jit --with-assertions
make -j$(nproc)
```

On 3.15.0b4 and main (9adef68a7956744b804d09c8b8b06c8c171053e5), this results in a `_bootstrap_python` that crashes on assertions:

```
$ make
./_bootstrap_python ./Programs/_freeze_module.py abc ./Lib/abc.py Python/frozen_modules/abc.h
_bootstrap_python: Python/pystate.c:925: interpreter_clear: Assertion `cold->vm_data.valid' failed.
make: *** [Makefile:1907: Python/frozen_modules/abc.h] Aborted (core dumped)
make: *** Deleting file 'Python/frozen_modules/abc.h'
```

For some files, I'm also/instead seeing:

```
_bootstrap_python: Python/pystate.c:933: interpreter_clear: Assertion `cold_dynamic->vm_data.valid' failed.
```

Bisect points to 4fa80ce74c6d9f5159bdc5ec3596a194f0391e21 which seems a huge change:

```
commit 4fa80ce74c6d9f5159bdc5ec3596a194f0391e21
Author:     Ken Jin <kenjin@python.org>
AuthorDate: 2025-11-13 19:08:32 +0100
Commit:     GitHub <noreply@github.com>
CommitDate: 2025-11-13 19:08:32 +0100

    gh-139109: A new tracing JIT compiler frontend for CPython (GH-140310)
```

CC @Fidget-Spinner 

### CPython versions tested on:

3.15, CPython main branch

### Operating systems tested on:

Linux

### Output from running 'python -VV' on the command line:

_No response_

<!-- gh-linked-prs -->
### Linked PRs
* gh-154142
* gh-154457
<!-- /gh-linked-prs -->

--- a/Lib/test/test_capi/test_opt.py
+++ b/Lib/test/test_capi/test_opt.py
@@ -5127,6 +5127,16 @@ class TestUopsOptimization(unittest.TestCase):
                               f" {executor} at offset {idx} rather"
                               f" than expected _EXIT_TRACE")
 
+    def test_jit_shutdown_after_cold_executor_creation(self):
+        script_helper.assert_python_ok("-c", textwrap.dedent(f"""
+            def f():
+                for x in range({TIER2_THRESHOLD + 3}):
+                    for y in range({TIER2_THRESHOLD + 3}):
+                        z = x + y
+
+            f()
+        """), PYTHON_JIT="1")
+
     def test_enter_executor_valid_op_arg(self):
         script_helper.assert_python_ok("-c", textwrap.dedent("""
             import sys

--- a/Misc/NEWS.d/next/Core_and_Builtins/2026-07-19-16-18-25.gh-issue-154014.CJsjVL.rst
+++ b/Misc/NEWS.d/next/Core_and_Builtins/2026-07-19-16-18-25.gh-issue-154014.CJsjVL.rst
@@ -1,0 +1,2 @@
+Fix a JIT assertion during interpreter shutdown by initializing ``vm_data``
+fields for cold executors that bypass ``_Py_ExecutorInit()``.

--- a/Python/optimizer.c
+++ b/Python/optimizer.c
@@ -1809,7 +1809,7 @@ make_cold_executor(uint16_t opcode)
     cold->jit_code = NULL;
     cold->jit_size = 0;
     if (_PyJIT_Compile(cold, cold->trace, 1)) {
-        Py_DECREF(cold);
+        PyExecutor_Free(cold);
         Py_FatalError("Cannot allocate core JIT code");
     }
 #endif

--- a/Python/optimizer.c
+++ b/Python/optimizer.c
@@ -1809,7 +1809,7 @@ make_cold_executor(uint16_t opcode)
     cold->jit_code = NULL;
     cold->jit_size = 0;
     if (_PyJIT_Compile(cold, cold->trace, 1)) {
-        PyExecutor_Free(cold);
+        _PyExecutor_Free(cold);
         Py_FatalError("Cannot allocate core JIT code");
     }
 #endif

--- a/Python/optimizer.c
+++ b/Python/optimizer.c
@@ -1788,7 +1788,6 @@ _Py_ExecutorInit(_PyExecutorObject *executor, const _PyBloomFilter *dependency_s
     }
     return 0;
 }
-
 static _PyExecutorObject *
 make_cold_executor(uint16_t opcode)
 {
@@ -1796,10 +1795,17 @@ make_cold_executor(uint16_t opcode)
     if (cold == NULL) {
         Py_FatalError("Cannot allocate core JIT code");
     }
+
     ((_PyUOpInstruction *)cold->trace)->opcode = opcode;
+
+    // Cold executors bypass _Py_ExecutorInit().
+    cold->vm_data.valid = true;
+    cold->vm_data.pending_deletion = 0;
+
     // This is initialized to false so we can prevent the executor
     // from being immediately detected as cold and invalidated.
     cold->vm_data.cold = false;
+
 #ifdef _Py_JIT
     cold->jit_code = NULL;
     cold->jit_size = 0;
@@ -1808,6 +1814,7 @@ make_cold_executor(uint16_t opcode)
         Py_FatalError("Cannot allocate core JIT code");
     }
 #endif
+
     _Py_SetImmortal((PyObject *)cold);
     return cold;
 }

--- a/Python/optimizer.c
+++ b/Python/optimizer.c
@@ -1800,6 +1800,7 @@ make_cold_executor(uint16_t opcode)
     // Cold executors bypass _Py_ExecutorInit().
     cold->vm_data.valid = true;
     cold->vm_data.pending_deletion = 0;
+    cold->vm_data.code = NULL;
 
     // This is initialized to false so we can prevent the executor
     // from being immediately detected as cold and invalidated.

--- a/Python/optimizer.c
+++ b/Python/optimizer.c
@@ -1788,6 +1788,7 @@ _Py_ExecutorInit(_PyExecutorObject *executor, const _PyBloomFilter *dependency_s
     }
     return 0;
 }
+
 static _PyExecutorObject *
 make_cold_executor(uint16_t opcode)
 {
@@ -1795,9 +1796,7 @@ make_cold_executor(uint16_t opcode)
     if (cold == NULL) {
         Py_FatalError("Cannot allocate core JIT code");
     }
-
     ((_PyUOpInstruction *)cold->trace)->opcode = opcode;
-
     // Cold executors bypass _Py_ExecutorInit().
     cold->vm_data.valid = true;
     cold->vm_data.pending_deletion = 0;
@@ -1805,7 +1804,6 @@ make_cold_executor(uint16_t opcode)
     // This is initialized to false so we can prevent the executor
     // from being immediately detected as cold and invalidated.
     cold->vm_data.cold = false;
-
 #ifdef _Py_JIT
     cold->jit_code = NULL;
     cold->jit_size = 0;
@@ -1814,7 +1812,6 @@ make_cold_executor(uint16_t opcode)
         Py_FatalError("Cannot allocate core JIT code");
     }
 #endif
-
     _Py_SetImmortal((PyObject *)cold);
     return cold;
 }

